File: scripts/day2-part1-optimized.py
import time

def is_repeated_twice(num):
    """Optimized version - same as simple for this problem"""
    s = str(num)
    length = len(s)
    
    if length % 2 != 0:
        return False
    
    half = length // 2
    first_half = s[:half]
    second_half = s[half:]
    
    return first_half == second_half

def solve_part1(input_data):
    ranges = [r.strip() for r in input_data.split(',') if r.strip()]
    total = 0
    
    for range_str in ranges:
        start, end = map(int, range_str.split('-'))
        
        for id_num in range(start, end + 1):
            if is_repeated_twice(id_num):
                total += id_num
    
    return total

# Read input
with open("../inputs/day-2-input.txt", "r") as f:
    input_data = f.read().strip()

# Benchmark
iterations = 1
start_time = time.perf_counter()

for _ in range(iterations):
    result = solve_part1(input_data)

end_time = time.perf_counter()
total_time = end_time - start_time
avg_time_us = (total_time / iterations) * 1_000_000

print(f"Python (Day 2 Part 1 - Optimized)")
print(f"Answer: {result}")
print(f"Average time: {avg_time_us:.3f} microseconds ({iterations} iterations)")
print(f"Total time: {total_time * 1000:.2f} ms")

File: scripts/day2-part1-optimized.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <chrono>
#include <sstream>

using namespace std;

bool isRepeatedTwice(long long num) {
    string s = to_string(num);
    int len = s.length();
    
    if (len % 2 != 0) return false;
    
    int half = len / 2;
    string firstHalf = s.substr(0, half);
    string secondHalf = s.substr(half);
    
    return firstHalf == secondHalf;
}

long long solvePart1(const string& input) {
    vector<pair<long long, long long>> ranges;
    stringstream ss(input);
    string range;
    
    while (getline(ss, range, ',')) {
        size_t pos = range.find('-');
        if (pos != string::npos) {
            long long start = stoll(range.substr(0, pos));
            long long end = stoll(range.substr(pos + 1));
            ranges.push_back({start, end});
        }
    }
    
    long long sum = 0;
    for (const auto& r : ranges) {
        for (long long id = r.first; id <= r.second; id++) {
            if (isRepeatedTwice(id)) {
                sum += id;
            }
        }
    }
    
    return sum;
}

int main() {
    // Read input
    ifstream file("../inputs/day-2-input.txt");
    string input;
    getline(file, input);
    file.close();
    
    // Benchmark
    const int iterations = 1;
    auto start = chrono::high_resolution_clock::now();
    
    long long result;
    for (int i = 0; i < iterations; i++) {
        result = solvePart1(input);
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::nanoseconds>(end - start);
    double avgMicros = duration.count() / (double)iterations / 1000.0;
    
    cout << "C++ (Day 2 Part 1 - Optimized)" << endl;
    cout << "Answer: " << result << endl;
    cout << "Average time: " << avgMicros << " microseconds (" << iterations << " iterations)" << endl;
    cout << "Total time: " << (duration.count() / 1000000.0) << " ms" << endl;
    
    return 0;
}
